packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js

```diff
@@ -1,61 +1,98 @@
 import { useEffect, useRef, useState } from 'react';
-import { useLocation } from 'react-router-dom';
+import { useHistory, useLocation } from 'react-router-dom';
 import { useApolloClient } from '@apollo/react-hooks';
 
 import addToCache from './addToCache';
 import getRouteComponent from './getRouteComponent';
 
+const CODE_PERMANENT_REDIRECT = 301;
+const CODE_TEMPORARY_REDIRECT = 302;
+const REDIRECT_CODES = [CODE_PERMANENT_REDIRECT, CODE_TEMPORARY_REDIRECT];
+
+const DEMO = new Map().set('/penelope-peasant-blouse.html', {
+    isRedirect: true,
+    relativeUrl: '/susanna-draped-tank.html'
+});
 const IS_LOADING = { isLoading: true };
 
 export const useMagentoRoute = () => {
-    const [componentMap, setComponentMap] = useState(new Map());
+    const [componentMap, setComponentMap] = useState(DEMO);
     const { apiBase } = useApolloClient();
+    const history = useHistory();
     const { pathname } = useLocation();
     const isMountedRef = useRef(false);
 
-    const routeData = componentMap.get(pathname);
+    // Keep track of whether we have been mounted yet.
+    // Note that we are not unmounted on page transitions.
+    useEffect(() => {
+        isMountedRef.current = true;
+
+        return () => {
+            isMountedRef.current = false;
+        };
+    }, []);
 
     // ask Magento for a RootComponent that matches the current pathname
     useEffect(() => {
-        // avoid asking if we already know the answer
-        const isKnown = componentMap.has(pathname);
+        // Avoid setting state if unmounted.
+        if (!isMountedRef.current) {
+            return;
+        }
 
-        // ask again following a prior failure
-        const isNotFound = isKnown && componentMap.get(pathname).id === -1;
-        const shouldReload = isNotFound && navigator.onLine;
+        const shouldFetch = data => {
```

Comment:
There's a lot going on in this effect, and your changes have definitely improved the readability. I would consider more simplification, and separating responsibilities. `shouldFetch` can probably be derived from state outside of the effect.

Reply:
Resolved in https://github.com/magento/pwa-studio/pull/2504/commits/5be73656c08286410d34942110d9546cf67f28b0.